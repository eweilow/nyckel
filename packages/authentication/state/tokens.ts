import CSRF from "csrf";

export const csrfTokensSingleton = new CSRF();
