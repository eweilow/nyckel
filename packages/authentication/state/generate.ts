import { CSRFPair } from "./pair";
import { csrfTokensSingleton } from "./tokens";
import CSRF from "csrf";

export async function generateCSRFPair(
  instance: CSRF = csrfTokensSingleton
): Promise<CSRFPair> {
  const secret = await instance.secret(); // The secret, stored in sessions
  const token = await instance.create(secret); // The state sent to Auth0

  return {
    secret,
    token
  };
}
