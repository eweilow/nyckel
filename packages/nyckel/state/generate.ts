import { CSRFPair } from "./pair";
import { csrfTokensSingleton } from "./tokens";

export async function generateCSRFPair(): Promise<CSRFPair> {
  const secret = await csrfTokensSingleton.secret(); // The secret, stored in sessions
  const token = await csrfTokensSingleton.create(secret); // The state sent to Auth0

  return {
    secret,
    token
  };
}
