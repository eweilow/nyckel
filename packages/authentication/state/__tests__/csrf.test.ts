import { generateCSRFPair } from "../generate";
import CSRF from "csrf";
import { csrfTokensSingleton } from "../tokens";
import { verifyCSRFPair } from "../..";

jest.mock("../tokens", () => {
  const CSRF = jest.requireActual("csrf");
  class Instance extends CSRF {
    constructor() {
      super();
    }

    secret = jest.fn(() => Promise.resolve("singleton:secret")) as any;
    create = jest.fn(() => Promise.resolve("singleton:token")) as any;
    verify = jest.fn(
      (secret, token) =>
        secret === "singleton:secret" && token === "singleton:token"
    ) as any;
  }
  return {
    csrfTokensSingleton: new Instance()
  };
});

describe("csrfTokensSingleton", () => {
  it("is instance of CSRF", () => {
    expect(csrfTokensSingleton).toBeInstanceOf(CSRF);
  });
});

describe("generateCSRFPair", () => {
  it("generates correctly", async () => {
    const generateSecret = jest.fn(() => Promise.resolve("secret"));
    const createToken = jest.fn(() => Promise.resolve("token"));

    const instance = {
      secret: generateSecret,
      create: createToken
    };

    expect(await generateCSRFPair(instance as any)).toEqual({
      secret: "secret",
      token: "token"
    });

    expect(generateSecret).toBeCalled();
    expect(createToken).toBeCalledWith(
      await generateSecret.mock.results[0].value
    );
  });

  it("generates correctly with default parameter", async () => {
    expect(await generateCSRFPair()).toEqual({
      secret: "singleton:secret",
      token: "singleton:token"
    });
    expect(csrfTokensSingleton.secret).toBeCalled();
    expect(csrfTokensSingleton.create).toBeCalledWith(
      await (csrfTokensSingleton.secret as jest.Mock).mock.results[0].value
    );
  });
});

describe("verifyCSRFPair", () => {
  it("throws on missing secret", async () => {
    await expect(
      verifyCSRFPair({ token: "token" } as any)
    ).rejects.toThrowErrorMatchingSnapshot();
  });
  it("throws on missing token", async () => {
    await expect(
      verifyCSRFPair({ secret: "secret" } as any)
    ).rejects.toThrowErrorMatchingSnapshot();
  });
  it("throws if unable to verify", async () => {
    await expect(
      verifyCSRFPair({ secret: "secret", token: "token" } as any)
    ).rejects.toThrowErrorMatchingSnapshot();
  });
  it("doesn't throw if it can verify", async () => {
    await expect(
      verifyCSRFPair({
        secret: "singleton:secret",
        token: "singleton:token"
      } as any)
    ).resolves.toBeUndefined();
  });
});
