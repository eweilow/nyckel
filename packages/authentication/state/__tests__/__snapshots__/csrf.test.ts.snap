// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verifyCSRFPair throws if unable to verify 1`] = `"The provided CSRF pair was not valid (token does not match secret)"`;

exports[`verifyCSRFPair throws on missing secret 1`] = `"Missing CSRF pair secret!"`;

exports[`verifyCSRFPair throws on missing token 1`] = `"Missing CSRF pair token!"`;
